File: src/main/java/viper/net/server/FileUploadChunkRelayEventListener.java
package viper.net.server;


import java.io.UnsupportedEncodingException;
import org.jboss.netty.buffer.ChannelBuffers;
import org.jboss.netty.channel.Channel;
import org.jboss.netty.channel.ChannelFutureListener;
import org.jboss.netty.handler.codec.http.DefaultHttpResponse;
import org.jboss.netty.handler.codec.http.HttpResponse;
import org.jboss.netty.handler.codec.http.HttpResponseStatus;
import org.jboss.netty.handler.codec.http.HttpVersion;
import org.json.JSONException;
import org.json.JSONObject;


public class FileUploadChunkRelayEventListener implements HttpChunkRelayEventListener
{
  public void onError(Channel clientChannel)
  {
    sendResponse(clientChannel, false);
  }

  public void onCompleted(Channel clientChannel)
  {
    sendResponse(clientChannel, true);
  }

  private void sendResponse(Channel clientChannel, boolean success)
  {
    try
    {
      HttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
      JSONObject jsonResponse = new JSONObject();
      jsonResponse.put("success", Boolean.toString(success));
      response.setContent(ChannelBuffers.wrappedBuffer(jsonResponse.toString(2).getBytes("UTF-8")));
      clientChannel.write(response).addListener(ChannelFutureListener.CLOSE);
    }
    catch (JSONException e)
    {
      e.printStackTrace();
    }
    catch (UnsupportedEncodingException e)
    {
      e.printStackTrace();
    }
  }
}

File: examples/src/main/scala/io/viper/examples/files/FileUploadChunkRelayEventListener.scala
package io.viper.examples.files


import io.viper.core.server.file.HttpChunkRelayEventListener
import io.viper.core.server.Util
import org.jboss.netty.handler.codec.http.{HttpResponseStatus, HttpVersion, DefaultHttpResponse}
import org.jboss.netty.buffer.ChannelBuffers
import org.json.{JSONException, JSONObject}
import java.io.UnsupportedEncodingException
import org.jboss.netty.channel.{Channel, ChannelFutureListener}
import java.util.{Map, UUID}


class FileUploadChunkRelayEventListener(hostname: String) extends HttpChunkRelayEventListener {
  def onError(clientChannel: Channel) {
    if (clientChannel != null) sendResponse(null, clientChannel, false)
  }

  def onCompleted(fileKey: String, clientChannel: Channel) = sendResponse(fileKey, clientChannel, true)

  def onStart(props: Map[String, String]) = Util.base64Encode(UUID.randomUUID())

  private def sendResponse(fileKey: String, clientChannel: Channel, success: Boolean) {
    try {
      val response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK)

      val jsonResponse = new JSONObject()
      jsonResponse.put("success", success)
      if (success) {
        jsonResponse.put("thumbnail", String.format("%s/thumb/%s", hostname, fileKey))
        jsonResponse.put("url", String.format("/d/%s", fileKey))
        jsonResponse.put("key", fileKey)
      }

      response.setContent(ChannelBuffers.wrappedBuffer(jsonResponse.toString(2).getBytes("UTF-8")))
      clientChannel.write(response).addListener(ChannelFutureListener.CLOSE)
    }
    catch {
      case e: JSONException => e.printStackTrace()
      case e: UnsupportedEncodingException => e.printStackTrace()
    }
  }
}
